### ISSUE TITLE
Endpoint panic, RC=137 during 1K IO 

### ISSUE THREAD

**Issue by akmithal** [noobaa/noobaa-core#6782]
### Environment info
- NooBaa Version: v4.9.0-202.ci
- Platform: OCP 4.8.9

### Actual behavior
1. IO caused endpoint panic 

### Expected behavior
1. IO shouldn't cause endpoint panic

### Steps to reproduce
1. Started upload of 1K objects and found panic on endpoint pod. Almost 100K objects were attempted. 

Endpoint pod restarted:

```
[root@ocp-akshat-1-inf ~]# podo
NAME                                               READY   STATUS    RESTARTS   AGE
noobaa-core-0                                      1/1     Running   0          3h
noobaa-db-pg-0                                     1/1     Running   0          3h8m
noobaa-default-backing-store-noobaa-pod-781468f4   1/1     Running   0          3h5m
noobaa-endpoint-866d5fc4b4-nnvqq                   1/1     Running   1          3h1m
noobaa-operator-784bb6685b-f9c52                   1/1     Running   0          3h13m
ocs-metrics-exporter-6cbf9c6bcb-7pn8q              1/1     Running   0          3h12m
ocs-operator-6c75d9cdb6-4k5cx                      1/1     Running   0          3h12m
odf-console-77dc4875d4-z6mkq                       1/1     Running   0          3h13m
odf-operator-controller-manager-568f657687-562g7   2/2     Running   1          3h13m
rook-ceph-operator-768c66d885-q77m7                1/1     Running   0          3h12m
```

```
Oct-26 7:52:33.729 [Endpoint/15]    [L0] core.endpoint.s3.ops.s3_put_object:: PUT OBJECT bucket-max max-objects/obj_75211  
Oct-26 7:52:33.730 [Endpoint/15]    [L0] core.sdk.object_sdk:: validate_non_nsfs_bucket:  { name: SENSITIVE-7fa8f881394f8a6d, email: SENSITIVE-3d288e97f8a66631, access_keys: [ { access_key: SENSITIVE-6b980914f9c02c55, secret_key: SENSITIVE-0c10683fd61ed392 } ], has_login: false, has_s3_access: true, allowed_buckets: { full_permission: true }, default_resource: 'nsfs-nsr-1', can_create_buckets: true, nsfs_account_config: { gid: 800, uid: 5000, nsfs_only: true, new_buckets_path: '/', backend: 'GPFS', warn_threshold_ms: 100 }, systems: [ { name: 'noobaa', roles: [ 'admin' ] } ], external_connections: { count: 0, connections: [] }, preferences: { ui_theme: 'LIGHT' } } { fs_root_path: '/nsfs/nsfs-nsr-1', fs_backend: 'GPFS', id: '617795c7f62ab4002ac4161a', name: 'nsfs-nsr-1' }
Oct-26 7:52:33.731 [Endpoint/15]    [L1] core.endpoint.s3.s3_rest:: S3 REQUEST PUT /bucket-max/max-objects/obj_75221 op put_object request_id kv7sjv7l-dusct6-xrq { 'accept-encoding': 'identity', 'user-agent': 'aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.19.1.el8_4.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.sync', 'content-md5': 'v2GerAzfP2jUluqTRBN+iw==', expect: '100-continue', 'x-amz-date': '20211026T075233Z', 'x-amz-content-sha256': 'UNSIGNED-PAYLOAD', authorization: 'AWS4-HMAC-SHA256 Credential=uGPZGM7U9HqzZPcHGufK/20211026/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=3f7c42da41a503b78292c0e692bf40931ddac32bb7f6ddeb89cb5be468604ba8', 'content-length': '512', host: 's3-openshift-storage.apps.ocp-akshat-1.cp.fyre.ibm.com', 'x-forwarded-host': 's3-openshift-storage.apps.ocp-akshat-1.cp.fyre.ibm.com', 'x-forwarded-port': '443', 'x-forwarded-proto': 'https', forwarded: 'for=10.17.66.193;host=s3-openshift-storage.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https', 'x-forwarded-for': '10.17.66.193' }
Oct-26 7:52:33.731 [Endpoint/15]    [L0] core.endpoint.s3.ops.s3_put_object:: PUT OBJECT bucket-max max-objects/obj_75221  
Oct-26 7:52:33.732 [Endpoint/15]    [L0] core.sdk.object_sdk:: validate_non_nsfs_bucket:  { name: SENSITIVE-7fa8f881394f8a6d, email: SENSITIVE-3d288e97f8a66631, access_keys: [ { access_key: SENSITIVE-6b980914f9c02c55, secret_key: SENSITIVE-0c10683fd61ed392 } ], has_login: false, has_s3_access: true, allowed_buckets: { full_permission: true }, default_resource: 'nsfs-nsr-1', can_create_buckets: true, nsfs_account_config: { gid: 800, uid: 5000, nsfs_only: true, new_buckets_path: '/', backend: 'GPFS', warn_threshold_ms: 100 }, systems: [ { name: 'noobaa', roles: [ 'admin' ] } ], external_connections: { count: 0, connections: [] }, preferences: { ui_theme: 'LIGHT' } } { fs_root_path: '/nsfs/nsfs-nsr-1', fs_backend: 'GPFS', id: '617795c7f62ab4002ac4161a', name: 'nsfs-nsr-1' }
Oct-26 7:52:33.732 [Endpoint/15]    [L1] core.endpoint.s3.s3_rest:: S3 REQUEST PUT /bucket-max/max-objects/obj_7521 op put_object request_id kv7sjv7n-dvvor1-xdb { 'accept-encoding': 'identity', 'user-agent': 'aws-cli/2.1.29 Python/3.8.8 Linux/4.18.0-305.19.1.el8_4.x86_64 exe/x86_64.rhel.8 prompt/off command/s3.sync', 'content-md5': 'v2GerAzfP2jUluqTRBN+iw==', expect: '100-continue', 'x-amz-date': '20211026T075233Z', 'x-amz-content-sha256': 'UNSIGNED-PAYLOAD', authorization: 'AWS4-HMAC-SHA256 Credential=uGPZGM7U9HqzZPcHGufK/20211026/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=a8b0fddf9eb7a16c25d7cb9307b8f4c3be0bb3d1daa484a6d3fa406b27e79325', 'content-length': '512', host: 's3-openshift-storage.apps.ocp-akshat-1.cp.fyre.ibm.com', 'x-forwarded-host': 's3-openshift-storage.apps.ocp-akshat-1.cp.fyre.ibm.com', 'x-forwarded-port': '443', 'x-forwarded-proto': 'https', forwarded: 'for=10.17.66.193;host=s3-openshift-storage.apps.ocp-akshat-1.cp.fyre.ibm.com;proto=https', 'x-forwarded-for': '10.17.66.193' }
Oct-26 7:52:33.732 [Endpoint/15]    [L0] core.endpoint.s3.ops.s3_put_object:: PUT OBJECT bucket-max max-objects/obj_7521  




######################################################################
/noobaa_init_files/noobaa_init.sh: line 75:    15 Killed                  $*
Tue Oct 26 07:52:33 UTC 2021 NooBaa: Process exited RIP (RC=137)
######################################################################
```

**Comment by akmithal** [comment]
[must-gather-0311_1.tar.gz](https://github.com/noobaa/noobaa-core/files/7465750/must-gather-0311_1.tar.gz)
[debug_9.zip](https://github.com/noobaa/noobaa-core/files/7465752/debug_9.zip)

I could reproduce it today while doing 100GB file upload run.

**Comment by akmithal** [comment]
[endpoint_rc_137_3Nov.zip](https://github.com/noobaa/noobaa-core/files/7465864/endpoint_rc_137_3Nov.zip)
@jeniawhite , here is the log for crashed endpoint log

**Comment by akmithal** [comment]
Me and @jeniawhite did the live debugging run and we realized the memory at the point of upload completion increased significantly for process "_node ./src/s3/s3rver_starter.js". 

Post this, the endpoint PANIC occurred. 

```
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         207  0.2  2.8 1394096 1385300 pts/2 Ss+  13:02   0:06 /bin/sh
root          14 33.4  1.3 1553388 677828 ?      Sl   12:11  29:01 node ./src/s3/s3rver_starter.js
root        1407  0.0  0.0  47592  3556 pts/4    R+   13:38   0:00 ps -aux --sort=-%mem
root         862  0.0  0.0  12056  3356 pts/4    Ss   13:27   0:00 /bin/sh
root         273  0.0  0.0  12056  3292 pts/3    Ss+  13:11   0:00 /bin/sh
root         103  0.0  0.0  12188  3080 pts/1    Ss+  12:47   0:00 /bin/sh
root          26  0.0  0.0  12184  2892 pts/0    Ss+  12:26   0:00 /bin/sh
root           1  0.0  0.0  12056  2716 ?        Ss   12:11   0:00 /bin/bash /noobaa_init_files/noobaa_init.sh init_endpoint
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         207  0.2  2.8 1394096 1385300 pts/2 Ss+  13:02   0:06 /bin/sh
root          14 33.4  1.3 1553388 677828 ?      Sl   12:11  29:01 node ./src/s3/s3rver_starter.js
root        1410  0.0  0.0  47592  3488 pts/4    R+   13:38   0:00 ps -aux --sort=-%mem
root         862  0.0  0.0  12056  3356 pts/4    Ss   13:27   0:00 /bin/sh
root         273  0.0  0.0  12056  3292 pts/3    Ss+  13:11   0:00 /bin/sh
root         103  0.0  0.0  12188  3080 pts/1    Ss+  12:47   0:00 /bin/sh
root          26  0.0  0.0  12184  2892 pts/0    Ss+  12:26   0:00 /bin/sh
root           1  0.0  0.0  12056  2716 ?        Ss   12:11   0:00 /bin/bash /noobaa_init_files/noobaa_init.sh init_endpoint
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         207  0.2  2.8 1394096 1385300 pts/2 Ss+  13:02   0:06 /bin/sh
root          14 33.4  2.7 2211392 1346504 ?     Rl   12:11  29:06 node ./src/s3/s3rver_starter.js
root        1413  0.0  0.0  47592  3556 pts/4    R+   13:38   0:00 ps -aux --sort=-%mem
root         862  0.0  0.0  12056  3356 pts/4    Ss   13:27   0:00 /bin/sh
root         273  0.0  0.0  12056  3292 pts/3    Ss+  13:11   0:00 /bin/sh
root         103  0.0  0.0  12188  3080 pts/1    Ss+  12:47   0:00 /bin/sh
root          26  0.0  0.0  12184  2892 pts/0    Ss+  12:26   0:00 /bin/sh
root           1  0.0  0.0  12056  2716 ?        Ss   12:11   0:00 /bin/bash /noobaa_init_files/noobaa_init.sh init_endpoint
```

**Comment by guymguym** [comment]
Hey @jeniawhite @akmithal 

First - did you try setting nodejs `--max-old-space-size` flag to make nodejs respect the same limits that we set to the pod? The reason I suspect that is that while we are limiting the buffers usage, we might need nodejs to also understand the memory limit and respect it by triggering GC at the right times.

In addition we can use heap snapshots to find out what is pilling up in memory:
1. make the `panic.memory_monitor()` interval a config.js variable so we can change in the pod spec and set to say 1 second. [ref](https://github.com/noobaa/noobaa-core/blob/f7ee1b2c4477464a3d050479ff7f3e8e58239b35/src/util/panic.js#L27)
2. call `panic.enable_heapdump()` from the endpoint main with relevant sizes such as next=1500 step=100 so that whenever the memory_monitor interval checks if the usage crossed 1500 MB it will write out a heapdump file, and then it will also write another one on 1600, and 1700, etc. [ref](https://github.com/noobaa/noobaa-core/blob/f7ee1b2c4477464a3d050479ff7f3e8e58239b35/src/util/panic.js#L29-L37)
3. BTW I see that in newer nodejs we can also use these flags (or with env `NODE_OPTIONS=--flag...`) to achieve the same heap dumps using `--heapsnapshot-near-heap-limit` and `--heap-prof`. [ref](https://nodejs.org/docs/latest-v14.x/api/cli.html#cli_heapsnapshot_near_heap_limit_max_count)
4. Set env `NOOBAA_INIT_MODE=auto` to make sure the container does not die and only the process restarts inside so we can download those heapdumps from the pod. [ref](https://github.com/noobaa/noobaa-core/blob/f7ee1b2c4477464a3d050479ff7f3e8e58239b35/src/deploy/NVA_build/noobaa_init.sh#L101-L105)
5. Use devtools to compare two heapdumps to detect which objects where added between those times - this should show you which objects are piling up in the memory, and also what is keeping them in memory. [ref](https://developer.chrome.com/docs/devtools/memory-problems/heap-snapshots/#comparison_view)

**Comment by akmithal** [comment]
Closing this defect as this fix is working with ODF 4.9 - 236.ci
